Unity3D/Assets/RosSharp/Editor/UrdfImporter/UrdfRobotExtensions.cs

```diff
@@ -59,5 +78,93 @@ public static void setKinematic(GameObject robot, bool isKinematic)
             foreach (Rigidbody rigidbody in rigidbodies)
                 rigidbody.isKinematic = isKinematic;
         }
+
+        public static bool ImportUrdfToUnity(Robot robot)
+        {
+            // Create URDF folder
+            if (!AssetDatabase.IsValidFolder("Assets/Urdf"))
+            {
+                AssetDatabase.CreateFolder("Assets", "Urdf");
+            }
+
+            // Create folder for URDF
+            if (!AssetDatabase.IsValidFolder("Assets/Urdf/" + robot.name))
+            {
+                AssetDatabase.CreateFolder("Assets/Urdf", robot.name);
+            }
+            // TODO: if folder exists, check if robot should be updated.
+
+            // Create folder for meshes
+            if (!AssetDatabase.IsValidFolder("Assets/Urdf/" + robot.name + "/meshes"))
+            {
+                AssetDatabase.CreateFolder("Assets/Urdf/" + robot.name, "meshes");
+            }
+
+            // Copy URDF into folder
+            string unity_urdf_file_path = Path.Combine(Application.dataPath + "/Urdf/" + robot.name, robot.name + ".urdf");
+            unity_urdf_file_path = unity_urdf_file_path.Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar);
+
+            // Find the root of the description package
+            string urdf_package_path = robot.filename;
+            bool package_dir_found = false;
+            do
+            {
+                // First loop, remove file name from string, next loops remove last directory from string
+                urdf_package_path = urdf_package_path.Remove(urdf_package_path.LastIndexOf(Path.AltDirectorySeparatorChar));
+                
+                if (File.Exists(Path.Combine(urdf_package_path, "package.xml")))
+                {
+                    package_dir_found = true;
+                    break;
+                }
+            } while (urdf_package_path.LastIndexOf(Path.AltDirectorySeparatorChar) != -1);
+
+            if (!package_dir_found)
+            {
+                // Ask user to manually locate the package path
+                if (EditorUtility.DisplayDialog("Import URDF Problem ",
+                    "Could not find the root of the robot_description package. Select folder?", "Yes", "Cancel Import"))
+                {
+                    urdf_package_path = EditorUtility.OpenFolderPanel("Select URDF root folder", "", "");
+                }
+                else
+                {
+                    // Abort the import process if user cancels.
+                    return false;
+                }
+            }
+            string package_name = urdf_package_path.Remove(0, urdf_package_path.LastIndexOf(Path.AltDirectorySeparatorChar) + 1);
+
+            // Load XML and copy mesh files
```

Comment:
The mesh import code (lines 138 -165) need major rework before inclusion into the master:
* First of all it is a new thing that should be put into a seprate method again.
* You let the urdf file getting read and search for mesh files befor it is actually read by the Robot constuctur. Reading a xml twice smells not good...
* Copying around multiple mesh files in the Asset folder is not good as well. When moving the urdf package e.g. due to a renaming of the robot, we suddenly have all meshes multiple times in the asset folder.
* Coping files when you actually want to import something is generally not nice.
* We definitely should not copy the urdf file such that two identical urdf files exist in our Asset folder.

Reply:
> You suggest reading the urdf file here and search for mesh files before it is actually being read for import by the Robot constuctur. Reading a file that you want to import twice smells not good...

I read the URDF then change the mesh filename attribute. I did this because UrdfAssetDatabase makes `assetPath` the location where the URDF is saved, then looks for the meshes in `assetPath` + `mesh filename from urdf`.  So if I have

`Assets/sawyer_description/sawyer.urdf` I have to put my meshes in `Assets/sawyer_description/sawyer_description/meshes` which I believe is a little confusing.

Maybe a change needs to be made to UrdfAssetDatabase?

> Copying around multiple mesh files in the Asset folder is not good as well. When moving the urdf package e.g. due to a renaming of the robot, we suddenly have all meshes multiple times in the asset folder.

The wiki asks users to copy the urdf and meshes manually from the robot description package. I don't see how this is different or understand why you say there are multiple meshes. Please elaborate on that part.

> Users do not expect extensive copying operations when they trigger an import. Copying files when you actually want to import something is generally not nice.

Same as above, the wiki asks users to copy files in manually. I fail to see the difference. 

> We definitely should not copy the urdf file such that two identical urdf files exist in our Asset folder.

Only one urdf file exists in the Assets folder. The URDF in the assets folder has the mesh `filename` attribute updated so UrdfAssetDatabase can find the mesh.